src/nvim/extmark.c

```diff
@@ -194,25 +194,32 @@ static bool extmark_setraw(buf_T *buf, uint64_t mark, int row, colnr_T col)
   return true;
 }
 
-/// Remove an extmark
+/// Remove an extmark pointed to by "itr" or from "ns_id" and "id"
 ///
-/// @return  0 on missing id
-bool extmark_del(buf_T *buf, uint32_t ns_id, uint32_t id)
+/// @return  line number of the deleted mark, -1 on missing id
+linenr_T extmark_del(buf_T *buf, MarkTreeIter *itr, MTKey key, uint32_t ns_id, uint32_t id)
 {
-  MarkTreeIter itr[1] = { 0 };
-  MTKey key = marktree_lookup_ns(buf->b_marktree, ns_id, id, false, itr);
-  if (!key.id) {
-    return false;
+  MarkTreeIter it[1] = { 0 };
+
+  if (itr == NULL) {
+    key = marktree_lookup_ns(buf->b_marktree, ns_id, id, false, it);
+    if (!key.id) {
+      return 0;
+    }
+    assert(key.pos.row >= 0);
+    itr = it;
   }
   assert(key.pos.row >= 0);
-  uint64_t other = marktree_del_itr(buf->b_marktree, itr, false);
 
   MTKey key2 = key;
-
+  uint64_t other = marktree_del_itr(buf->b_marktree, itr, false);
   if (other) {
     key2 = marktree_lookup(buf->b_marktree, other, itr);
     assert(key2.pos.row >= 0);
     marktree_del_itr(buf->b_marktree, itr, false);
+    if (id == 0) {
+      marktree_itr_get(buf->b_marktree, key.pos.row, key.pos.col, itr);
```

Comment:
https://github.com/neovim/neovim/pull/25724#discussion_r1378225160 for reference, feel free to close if it doesn't make sense.

Reply:
So is it expected that when deleting during iteration a `marktree_itr_get()` is necessary to "restore" the iterator to a mark in the deleted position (if available)? I failed to reason about.
Also expected it might be possible to ensure this isn't necessary by skipping over `mt_end()` marks at the `extmark_del()` call site but it doesn't seem to work.

Reply:
ok, first what is happening here is that you move the argument iterator `itr` from the start mark to the position of the end mark. But even if we used a separate iterator `itr2` just for the end mark, it might restructure the marktree in a way so that `itr` becomes invalid anyway (we might be able to detect this more specifically in the future, but for now we just assume that any iterator becomes invalid). 

So for now, what you do here is the correct workaround.